runtime/src/accounts_db.rs

```diff
@@ -7124,4 +7260,302 @@ pub mod tests {
+                // to storage, should only contain one account
+                assert_eq!(
+                    slot_accounts[0],
+                    std::iter::once(keys[*slot as usize])
+                        .into_iter()
+                        .collect::<HashSet<Pubkey>>()
+                );
+            }
+        }
+    }
+
+    #[test]
+    fn test_accounts_db_cache_clean_max_root() {
+        let max_clean_root = 5;
+        run_test_accounts_db_cache_clean_max_root(10, max_clean_root);
+        // Test that if there are > MAX_CACHE_SLOTS in the cache after flush, then more roots
+        // will be flushed
+        run_test_accounts_db_cache_clean_max_root(
+            MAX_CACHE_SLOTS + max_clean_root as usize + 2,
+            max_clean_root,
+        );
+    }
+
+    fn run_flush_rooted_accounts_cache(should_clean: bool) {
+        let num_slots = 10;
+        let (accounts_db, keys, slots) = setup_accounts_db_cache_clean(num_slots);
+        let mut cleaned_bytes = 0;
+        let mut cleaned_accounts = 0;
+        let should_clean_tracker = if should_clean {
+            Some((&mut cleaned_bytes, &mut cleaned_accounts))
+        } else {
+            None
+        };
+
+        // If no cleaning is specified, then flush everything
+        accounts_db.flush_rooted_accounts_cache(None, should_clean_tracker);
+        for slot in &slots {
+            let slot_accounts = accounts_db.scan_account_storage(
+                *slot as Slot,
+                |loaded_account, _, slot_account: &mut HashSet<Pubkey>| {
+                    slot_account.insert(*loaded_account.pubkey());
+                    if let LoadedAccount::Cached(_) = loaded_account {
+                        panic!("All roots should have been flushed to storage");
+                    }
+                },
+            );
+            if !should_clean || slot == slots.last().unwrap() {
+                // The slot was not cleaned before being flushed to storage,
+                // so it also contains all the original updates.
+                assert_eq!(
+                    slot_accounts[0],
+                    keys[*slot as usize..]
+                        .iter()
+                        .cloned()
+                        .collect::<HashSet<Pubkey>>()
+                );
+            } else {
+                // If clean was specified, only the latest slot should have all the updates.
+                // All these other slots have been cleaned before flush
+                assert_eq!(
+                    slot_accounts[0],
+                    std::iter::once(keys[*slot as usize])
+                        .into_iter()
+                        .collect::<HashSet<Pubkey>>()
+                );
+            }
+        }
+    }
+
+    #[test]
+    fn test_flush_rooted_accounts_cache() {
+        run_flush_rooted_accounts_cache(true);
+        run_flush_rooted_accounts_cache(false);
```

Comment:
btw, I like to split these tests like this to see separated individual test results easily:

https://github.com/solana-labs/solana/pull/14557/files#diff-281d947e593a76f3c3380804162dc8cf426fcc33e0ae1192589c8193221cd1e2R3261